src/langchain_google_memorystore_redis/vector_store.py

```diff
@@ -0,0 +1,789 @@
+            query (str): The text query based on which similar documents are to be retrieved.
+            k (int): Specifies the number of documents to return, effectively setting a limit
+                on the size of the result set. Defaults to 4.
+            **kwargs (Any): A flexible argument allowing for the inclusion of additional
+                search parameters or options.
+
+        Returns:
+            List[Document]: A list containing up to 'k' Document objects, ranked by their
+                similarity to the query. These documents represent the most relevant results
+                found by the search operation, subject to the additional constraints and
+                configurations specified.
+
+        Raises:
+            ValueError: If any of the provided search parameters are invalid or if the search
+                operation encounters an error due to misconfiguration or execution issues within
+                the search backend.
+        """
+        # Embed the query using the embedding function
+        query_embedding = self.embeddings_service.embed_query(query)
+        return self.similarity_search_by_vector(query_embedding, k, **kwargs)
+
+    def max_marginal_relevance_search(
+        self,
+        query: str,
+        k: int = 4,
+        fetch_k: int = 20,
+        lambda_mult: float = 0.5,
+        **kwargs: Any,
+    ) -> List[Document]:
+        """
+        Performs a search to find documents that are both relevant to the query and diverse
+        among each other based on Maximal Marginal Relevance (MMR).
+
+        The MMR algorithm optimizes a combination of relevance to the query and diversity
+        among the results, controlled by the lambda_mult parameter.
+
+        Args:
+            query (str): The query string used to find similar documents.
+            k (int): The number of documents to return.
+            fetch_k (int): The number of documents to fetch for consideration. This should
+                be larger than k to allow for diversity calculation.
+            lambda_mult (float): Controls the trade-off between relevance and diversity.
+                Ranges from 0 (max diversity) to 1 (max relevance).
+            **kwargs: Additional keyword arguments
+
+        Returns:
+            List[Document]: A list of document objects selected based on maximal marginal relevance.
+
+        Raises:
+            ValueError: If lambda_mult is not in the range [0, 1].
+        """
+        # Validate the lambda_mult parameter to ensure it's within the valid range.
+        if not 0 <= lambda_mult <= 1:
+            raise ValueError("lambda_mult must be between 0 and 1.")
+
+        # Embed the query using a hypothetical method to convert text to vector.
+        query_embedding = self.embeddings_service.embed_query(query)
+
+        # Fetch initial documents based on query embedding.
+        initial_results = self._similarity_search_by_vector_with_score_and_embeddings(
```

Comment:
You should be able to use this function to implement this https://github.com/langchain-ai/langchain/blob/20a56fe0a2f1bb46dba9f70b9e94ebb64002f6d9/libs/community/langchain_community/vectorstores/utils.py#L23

Reply:
Can you elaborate more on this? We do use maximal_marginal_relevance function in L785. CC @PingXie

Reply:
Oops, I see it now. lgtm